Support generic lifetime arguments in method calls

petrochenkov opened the pull request:
Fixes https://github.com/rust-lang/rust/issues/42403
Fixes https://github.com/rust-lang/rust/issues/42115
Lifetimes in a method call `x.f::<'a, 'b, T, U>()` are treated exactly like lifetimes in the equivalent UFCS call `X::f::<'a, 'b, T, U>`.
In addition, if the method has late bound lifetime parameters (explicit or implicit), then explicitly specifying lifetime arguments is not permitted (guarded by a compatibility lint).
[breaking-change] because previously lifetimes in method calls were accepted unconditionally.

r? @eddyb

eddyb commented:
cc @nikomatsakis Whom I've discussed this with previously.

So I'm not sure I still have them somewhere, but I've done a crater run on a stricter version of this (just banning lifetimes) and found a few problems, mostly related to the fact that late-bound lifetimes do not appear in `ty::Generics`. I'm not even sure it's a good idea to expose the distinction.

There were some ideas to prevent explicit lifetimes when any ended up being late-bound, or outright support them (after #42417 we could start playing with that idea).

nikomatsakis commented:
So I am pretty nervous about exposing *only* the early-bound lifetimes. This ties in to that other issue, where we *do* allow such regions sometimes. I really would prefer it we only allow users to manually specify regions when they are all early-bound, so we can do it realiably and in a clearly forwards compatible way.

petrochenkov commented:
>only allow users to manually specify regions when they are all early-bound, so we can do it realiably and in a clearly forwards compatible way

I'll implement this then for both method calls and UFCS as a deny-by-default lint for cratering.

nikomatsakis commented:
Discussed in the @rust-lang/lang meeting -- we agree that restricting to 100% early-bound is the right thing here.

alexcrichton commented:
ping @petrochenkov, just wanted to keep this on your radar!

bors commented:
:umbrella: The latest upstream changes (presumably #42856) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
I've almost submitted an update implementing https://github.com/rust-lang/rust/pull/42492#issuecomment-307556752, but then realized that late bound lifetime parameters can be defined implicitly through lifetime elision (e.g. `fn f(_: &u8) {}`).
@eddyb Is there an easy way to detect presence of such implicit parameters during type checking? I haven't investigated yet.

eddyb commented:
You can detect them but I wouldn't consider them parameters for the purposes of paths, as I wouldn't an `impl Trait` somewhere in an argument type.

petrochenkov commented:
@eddyb 
>I wouldn't consider them parameters for the purposes of paths

It's just strange that these two functions are identical
```rust
fn late_explicit<'early, 'late>(_: &'late u8) -> &'early u8 { loop {} }
fn late_implicit<'early       >(_: &      u8) -> &'early u8 { loop {} }
```
, but `late_implicit::<'static>` would be legal and `late_explicit::<'static>` would be not. Writing out the late bound lifetime explicitly would be a breaking change then.

eddyb commented:
IMO there is no other real choice if we want the explicitly written parameters to mean *anything*.

petrochenkov commented:
>if we want the explicitly written parameters to mean anything.

I'm not sure what do you mean.

(I'm still going to check for implicit parameters, at least to look at what crater will say.)

eddyb commented:
@petrochenkov I meant that I don't think we can reasonably let you write anything *other* than what is present between `<...>` syntactically in the definition of the item, in a path to that item.

But I suppose I'm okay with going the even more conservative route of banning explicit parameters of a certain kind when there are implicit ones of that kind which *could* be later made explicit.

petrochenkov commented:
I believe I've extracted all the necessary logic to catch late bound parameters from `resolve_lifetime.rs`.
This should be ready for crater now.

(There's also an interesting case with lifetimes in struct/variant constructors which are currently treated as early bound (https://github.com/rust-lang/rust/issues/30904), but I haven't touched it, only added a test.)

eddyb commented:
Started crater run.

eddyb commented:
Crater run [failed](https://taskcluster.github.io/unified-logviewer/?url=https%3A%2F%2Fqueue.taskcluster.net%2Fv1%2Ftask%2F0z6ADwXfRy-szrN4mJ-emw%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log&highlightStart=7669&highlightEnd=7669&lineNumber=7669&wrapLines=false&showLineNumbers=true&jumpToHighlight=true&followLog=false&lineHeight=19&asText=false) due to the inclusion of `rustfmt` in the build.

cc @rust-lang/dev-tools

petrochenkov commented:
>Crater run failed due to the inclusion of rustfmt in the build.

@nikomatsakis used crater [recently](https://github.com/rust-lang/rust/pull/42894#issuecomment-311430059) and everything seems working again.

eddyb commented:
@petrochenkov No, I mean, your changes broke `rustfmt` (if you look, the errors are about AST changes you made).

petrochenkov commented:
Oh, right, didn't notice, sorry for the noise.

Mark-Simulacrum commented:
I don't think this is waiting on team, so removing -- perhaps added by accident?

petrochenkov commented:
@Mark-Simulacrum 
I've added the label because @rust-lang/dev-tools  was pinged (https://github.com/rust-lang/rust/pull/42492#issuecomment-310897984) regarding rustfmt build preventing crater run.

alexcrichton commented:
ping @nrc, do you have suggestions about how to resolve the error on crater above?

eddyb commented:
Only now did it cross my mind that the libsyntax AST changes can be accepted separately, and that will unbreak the rustfmt build. @petrochenkov Mind splitting those into a new PR?
You will need to update rustfmt & the RLS submodule in that same PR but at least you know it will get merged right away (r=me on it, assuming it causes no breaking changes for users).

petrochenkov commented:
@eddyb 
Will do.

To be sure, PR merges to https://github.com/rust-lang-nursery/rustfmt aren't blocked on CI, right?
Technically, my PR is going to break rustfmt's master for a couple of days.

nrc commented:
> ping @nrc, do you have suggestions about how to resolve the error on crater above?

huh, this is much more complex than I anticipated. We don't yet include rustfmt directly, but because it is used by the RLS we have a compilation failure. I would suggest that Crater not build tools it doesn't need to as the best fix here (no need for Crater to build the RLS, aiui). I'll merge the fix to rustfmt and then pull it in to the RLS (on branches), then to land this PR (or the one with the libsyntax breaking changes), we'll need to update the RLS submodule to the RLS branch.

nrc commented:
To merge, you'll need to update the RLS submodule to https://github.com/rust-lang-nursery/rls/tree/rustfmt-42492, hopefully then everything should work.

petrochenkov commented:
I've submitted https://github.com/rust-lang/rust/pull/43115 with AST/HIR data changes and RLS update.

bors commented:
:umbrella: The latest upstream changes (presumably #43115) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
@eddyb 
Rebased, should be ready for crater now.

eddyb commented:
Started crater builds.

eddyb commented:
[Crater report](https://gist.github.com/eddyb/4fcf299cfa45c8945c155c4089267f33) shows some lifetime inference regressions but most of them are... type inference?
I'm not sure exactly what's happening there.

petrochenkov commented:
Looks like some indexes for parameters in `instantiate_method_substs` are incorrectly shifted.
Will try to fix this today.

petrochenkov commented:
Updated with a fix for the failures from crater.

eddyb commented:
Starting a second crater run.

eddyb commented:
[Second crater report](https://gist.github.com/eddyb/dfb714f0884ff8e74e430216dfa6e59a) shows fewer regressions, only 2 of which are due to late bound lifetimes.

eddyb commented on src/librustc_typeck/collect.rs:
```diff
@@ -772,6 +772,92 @@ fn trait_def<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                    -> bool {
+    struct LateBoundRegionsDetector<'a, 'tcx: 'a> {
+        tcx: TyCtxt<'a, 'tcx, 'tcx>,
+        binder_depth: usize,
```
This isn't used - the intention was probably to avoid false positives from e.g. `for<'a> fn(&'a T)`, right?

eddyb commented on src/librustc_typeck/collect.rs:
```diff
@@ -772,6 +772,92 @@ fn trait_def<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+
+            match self.tcx.named_region_map.defs.get(&lt.id).cloned() {
+                Some(rl::Region::Static) | Some(rl::Region::EarlyBound(..)) => {}
+                _ => self.has_late_bound_regions = true
```
This should compare the `binder_depth` to the depth in `rl::Region`, and perhaps not count anonymous regions, but I'm fine with the more conservative approach of including them.
`fn foo(_: for<'a> fn(&'a T)) {}` is the main false positive to avoid here.

petrochenkov replied:
Wait, I clearly remember writing the `binder_depth` comparison, I've even added tests to make sure it works (`early_tricky_explicit`/`early_tricky_implicit`), they shouldn't pass without it.
This is some mystery, I'll investigate.

petrochenkov replied:
Okay, the tests pass because the errors are reported as lints at later stage which doesn't happen, they fail if extracted into a separate file. I'll push the fix tomorrow, it's too late now.

petrochenkov commented:
@eddyb 
Updated with a fix for https://github.com/rust-lang/rust/pull/42492#discussion_r126916646
I also fixed the regression in `roaring-0.4.2` found by the second crater run (e.g. https://tools.taskcluster.net/groups/6khR64OzRhiJiTHGJSuiiA/tasks/6khR64OzRhiJiTHGJSuiiA/runs/0/logs/public%2Flogs%2Flive.log) where a type inference error is reported instead of the `late_bound_lifetime_arguments` lint.
I've left the lint to be deny-by-default given the tiny number of regressions.

eddyb commented:
@petrochenkov Started another crater run.

eddyb commented:
[Third crater report](https://gist.github.com/eddyb/8e438453185125b3a3e43dfaeef40ac5) shows 4 regressions.
Since the information was never used, I'm not sure why the authors kept the lifetimes (but really, it's our fault for parsing and ignoring them, some of them could be just "documentation").
There is at least one regression in `unsafe` code, which could *potentially* mean an invariant wasn't really enforced, but function signatures should prevent a real problem from spreading.

Mark-Simulacrum commented:
I guess this is waiting on review? Or perhaps a decision from the lang team?

nikomatsakis commented:
I think the relevant *decisions* have all been made -- we want to limit explicit lifetimes to cases that have only early-bound lifetimes for now (which are clearly backwards compatible). However, I'm not sure of the current back-and-forth between @eddyb and @petrochenkov on the state of the PR itself, it seems like they are iterating on crater runs. 

As an aside, @brson has been complaining to me that people still use crater, and suggesting that we should try to do some cargobomb runs for better coverage. @brson-- who should be ping to try and get that going?

petrochenkov commented:
@nikomatsakis 
>However, I'm not sure of the current back-and-forth between @eddyb and @petrochenkov on the state of the PR itself, it seems like they are iterating on crater runs.

The iteration is complete, all found regressions and mistakes are fixed.

nikomatsakis commented on src/librustc/lint/builtin.rs:
```diff
@@ -205,6 +205,12 @@ declare_lint! {
 
 declare_lint! {
+    pub LATE_BOUND_LIFETIME_ARGUMENTS,
+    Deny,
```
Do we really want to default to `Deny` here?

petrochenkov commented:
The lint level is changed to warn-by-default.
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 39114f9 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 39114f916905b16b414f51031426309b63f856a9 -->

bors commented:
:hourglass: Testing commit 39114f916905b16b414f51031426309b63f856a9 with merge 7c758abdad01aaa219c2cb14e4ed6878fb69c236...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/254692424?utm_source=github_status&utm_medium=notification)

alexcrichton commented:
@bors: retry

* curl failure

bors commented:
:hourglass: Testing commit 39114f916905b16b414f51031426309b63f856a9 with merge 29d60228266412b6bee73098d00ef185d781a8f3...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/254701185?utm_source=github_status&utm_medium=notification)

est31 commented:
@bors retry

bors commented:
:hourglass: Testing commit 39114f916905b16b414f51031426309b63f856a9 with merge 83c659ef655b1f740777f83eb415fd7ebe5a3fe5...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3958), [status-travis](https://travis-ci.org/rust-lang/rust/builds/254798184?utm_source=github_status&utm_medium=notification)
Approved by: nikomatsakis
Pushing 83c659ef655b1f740777f83eb415fd7ebe5a3fe5 to master...
